packages/api/src/util/animal.js

```diff
@@ -33,3 +38,101 @@ export const assignInternalIdentifiers = async (records, kind) => {
     }),
   );
 };
+
+/**
+ * Asynchronously checks if the given animal or batch has a type or breed already stored in the database.
+ * If not, it adds the type and/or breed to the database, updates the corresponding IDs, and removes
+ * the type_name or breed_name properties from the animal or batch object.
+ *
+ * @param {Object} req - The request object, containing the body with type and breed maps.
+ * @param {Object} animalOrBatch - The animal or batch object that contains type_name or breed_name properties.
+ * @param {number} farm_id - The ID of the farm to associate with the type or breed.
+ * @param {Object} trx - A transaction object for performing the database operations within a transaction.
+ *
+ * @returns {Promise<void>} - A promise that resolves when the type and breed IDs have been added/updated and the object has been modified.
+ *
+ * @throws {Error} - If any database operation fails.
+ */
+export const checkAndAddCustomTypeAndBreed = async (req, animalOrBatch, farm_id, trx) => {
```

Comment:
Your totally right, I considered adding it to the model I forget why I didn't... maybe because it would double the code. 

I think we came to some agreement a while ago to prefer fat models over fat controllers. I think I like the idea of services because I like lean controllers, and lean models as just a structured description of the db -- but we don't have that services construct yet.

Happy to add a new folder called services or something! But it might be good to get buy in from team. @kathyavini & @SayakaOno